### PR TITLE
fix(merge:branches): Merge source into target branch instead of the other way

### PR DESCRIPTION
**Pull request by tylerbutler**
In the existing logic and recommended steps for developers to follow, we merge `next` into `main`. While the resulting merge commit is the same regardless of direction, the direction _does_ matter when using tools to review history, especially visual tools (e.g. GitKraken, SourceTree, etc.). See [this StackOverflow answer](https://stackoverflow.com/a/45973607/551579) for details.

I updated the logic we use to change the direction of the merge. That will fix the PRs with no conflicts. For conflicting PRs, developers have to do it themselves I updated the instructions we include with more details about what to do. They need to check out `next`, then merge that with the temporary merge branch, then checkout the merge branch again and fast-forward it to the merge commit. That is:

- `git checkout -b temp-next upstream/next` -- make a temporary branch.
- `git merge upstream/main-next-abc1234` -- merge the main commits into the next branch.
- Resolve any merge conflicts between the branches, then commit all the changes
- `git checkout main-next-abc1234` -- check out the mergeBranch locally
- `git merge temp-next --ff-only` -- fast-forward to the merge commit
- `git push upstream` -- and push it to upstream

This is more complicated, but it's necessary to make our commit history legible.

### CODE REVIEW

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 1]
```diff
@@ -100,12 +101,14 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
-
 		if (this.gitRepo === undefined) {
 			this.errorLog(`gitRepo undefined: ${JSON.stringify(this.gitRepo)}`);
 			this.error("gitRepo is undefined", { exit: 1 });
 		}
 
+		// const remote =
+		// 	flags.remote ?? (await this.gitRepo.getRemote(context.originRemotePartialUrl));
+
 		const [owner, repo] = context.originRemotePartialUrl.split("/");
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:112]:
```suggestion
		const [owner, repo] = context.originRemotePartialUrl.split("/");
```

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 2]
```diff
@@ -10,6 +10,8 @@ import { strict as assert } from "node:assert";
 import { BaseCommand } from "../../base";
 import { Repository, createPullRequest, getCommitInfo, pullRequestExists } from "../../lib";
 
+const DEFAULT_REMOTE = "origin";
+
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:14]:
```suggestion
```

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 3]
```diff
@@ -114,7 +117,10 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 
 		this.remote =
 			flags.remote ?? (await this.gitRepo.getRemote(context.originRemotePartialUrl));
-		this.verbose(`Remote is: ${this.remote}`);
+		this.warning(`Remote is: ${this.remote}`);
+		if (this.remote === undefined) {
+			this.error("Remote for upstream repo not found", { exit: 1 });
+		}
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:123]:
```suggestion
		if (this.remote === undefined) {
			this.error("Remote for upstream repo not found", { exit: 1 });
		}
		this.info(`Remote is: ${this.remote}`);
```

**Comment by sonalideshpandemsft** [comment]
when is the `temp-next` branch deleted?

**Review by sonalideshpandemsft — APPROVED** [review]

**Comment by tylerbutler** [comment]
> when is the `temp-next` branch deleted? even though temp-next is not pushed to remote, should we update the description to say `Delete temp-next`?

At the end of the run method we call: `await this.doCleanup();` which deletes any branches marked for cleanup. We mark the temp branches for cleanup right after the variables are created. I could probably remove the other explicit deletion because it could just happen in cleanup, but I'll leave that for later if needed.

Adding it to the description is a good idea. Will do.

**Comment by sonalideshpandemsft** [comment]
one general question: Are folks allowed to use GitHub UI for resolving merge conflicts instead of pulling the branch locally? 

There are multiple times when we have to clean-up the PR one of which is when GitHub UI is used, and the commit message is `Merge next into next`. Should we probably add a note saying please don't use the UI? any thoughts around this?

**Pull request merged**